### PR TITLE
SupplyTrackerV3

### PR DESCRIPTION
**Pull request by Dyldozer**
Automatically keeps track of all consumables, ammo, trident casts, teleport tablets, and teleport scrolls used during the session and allows user to add, edit, or delete the amount.(Additions are right click submenu, Amount edit is pulls up a new gui element)
Works for all consumables, ammo, throwing weapons(including chins), trident casts(seas, swamp, sang, and their (e) versions), teleport scrolls, and teleport tablets.
Works with tick eating.
Built in protection against spam clicking adding too many of one item.
Arrow tracking works even if ammo type is switched or more of the same type is added.

[Main Features](https://i.gyazo.com/bd1d02ac3522de673ce0a373be9f52e0.mp4)
[Longer session test](https://i.gyazo.com/0d2e3a5dce47a7dedc6193f08ab3929f.png)
[Arrow tracking](https://i.gyazo.com/a9ae578f414005268af5fbd264bb2fd7.mp4)
[Throwing weapon Tracking](https://i.gyazo.com/1cd2ac25dd03bb1cf0d19694b6ff37f4.mp4)
[Add items from sub-menu](https://i.gyazo.com/f955f90cada806ff54a0341b1b3febc2.mp4)
[10 trip Zulrah test](https://i.gyazo.com/31492212d278f356502ceb7bf031e0a8.png)

Took a break from Runescape that's why I never finished it last time but I believe I made the changes requested(using ItemID instead of hardcoding ids and generating the menu from an enum) and added the new items and moved the calculations off swing thread.

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/suppliestracker/SuppliesTrackerPanel.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
- /*
+	/*
```
Comment (Spedwards) [inline runelite-client/src/main/java/net/runelite/client/plugins/suppliestracker/SuppliesTrackerPanel.java]: Remove this tab and the one in SuppliesTrackerPlugin.

Comment (Dyldozer) [reply runelite-client/src/main/java/net/runelite/client/plugins/suppliestracker/SuppliesTrackerPanel.java]: Thank you finished

**Comment by Beagon** [comment]
The idea is really nice. I tested this out for 15 minutes and will keep testing this for a little while. What I noticed:

- Doesn't track cannonballs being used. Would be really nice to have for slayer tracking.
- The UI is pretty ugly and doesn't fit RuneLite at all, maybe you can implement the UI that the Loot Tracker uses?

Still very good idea, but still needs some work.

**Comment by Dyldozer** [comment]
Thank you for the feedback.  I'll get started on the cannon tracking. The ui's base(background, font,   and overall panel at the top) is a copy of loot trackers. What would you like to see different within the ui or how do you think I should change to make it more aligned with current ui elements?

**Comment by Beagon** [comment]
> 
> 
> Thank you for the feedback. I'll get started on the cannon tracking. The ui's base(background, font, and overall panel at the top) is a copy of loot trackers. What would you like to see different within the ui or how do you think I should change to make it more aligned with current ui elements?

No problem at all, thanks for being so quick to reply and accepting my feedback 👍 

My suggestion for the UI:
- Put everything in a grid like the loot tracker does. Here is a rough mock-up of what I think it could look like, the red dot represent a mouse hovering over the item.
Mockup

Another bug/issue thing I found:
When I take a sip of a potion, in the tracker it counts it as 1 dose. The problem is, the prices between doses sometimes are really big. With this I mean, sometimes a 1 dose of Super Restore costs 4K while a 4 dose costs 10K, which makes the supplies used inaccurate. Certainly if you track a whole session.

Solutions:
- Make doses stack up. So if you take 1 sip it will become 1 dose used, 2 sips it will change to a 2 dose potion, 3 sips it will change to a 3 dose potion until it's at 4 (full) and then it will stay at 4 and add another 1 dose the next sip.
- Count a sip as 1/4th of the cost of a full potion instead of the price of a 1 dose version.

Both solutions should have a way of turning it on/off in the config of course but be on by default.

Another suggestion I came up with but isn't necessary for the MVP (Minimal viable plugin):
- Lap Function (Like you have on a stopwatch).
If you go with the Grid UI, you could make an option to right click on the grid title (total supplies, cost) and get the option to "Lap" that entry. This would create a new grid where all the supplies after lapping are tracked. This way people can try and see each trip/each kill what supplies they used if they want to optimize their trips. I hope you get the gist of what I mean, otherwise I'll try to explain it a little better.

Thanks again for the work and if you want me to help out code something or w/e. Let me know and I see  if I can find some time this week.

**Comment by Dyldozer** [comment]
Ahhh I see what you mean on the ui yeah that can be changed that's a really good idea! I'll have to recheck the potions again it should be set to if its a potion the value is the cost of a 4 dose/4 since I figured most people would be buying 4 doses. And I am actually working on a trip counter similar to your lap idea its just taking awhile to redo it to a separate object but I really do like the idea. 

Again thank you for the feedback and it being so well thought out and constructive. Any other ideas or changes you think would look better I would definitely love to hear!

**Comment by Beagon** [comment]
> 
> 
>  I'll have to recheck the potions again it should be set to if its a potion the value is the cost of a 4 dose/4 since I figured most people would be buying 4 doses.

Ah, I didn't figure that out, haven't really paid much attention to the code yet. Was more foccused on the end result. It might be better to be able to toggle that off/on. Also it would be great if you could make it so that it's actually clear that it's using the 4 dose prices. Because I saw 1 dose potions in the list I thought it counted as a 1 dose potion.


>
>
> Again thank you for the feedback and it being so well thought out and constructive.

No problem at all! This was an idea I had in my head myself but just didn't get around to build so when I saw it in the pull request list I had to check it out and give some feedback/insights of what I had in my head ;)

Keep me posted on any updates and I'll be sure to test them ASAP 👍

**Comment by leele2** [comment]
Love this idea, Noticed it doesn't track Blowpipe and thought maybe it could track coin purchases ( such as charter when going to Zulrah ) Hope to see this in the live version some time soon

**Comment by leele2** [comment]
Also could add enchanted jewellery teleports and rings of recoil

**Comment by james-munson** [comment]
This causes lag in and out at Zulrah 


`2019-03-10 22:24:31 [AWT-EventQueue-0] ERROR net.runelite.client.RuneLite - Uncaught exception:
java.util.ConcurrentModificationException: null
        at java.util.HashMap$HashIterator.nextNode(Unknown Source)
        at java.util.HashMap$KeyIterator.next(Unknown Source)
        at net.runelite.client.plugins.suppliestracker.SuppliesTrackerPanel.addRow(SuppliesTrackerPanel.java:266)
        at net.runelite.client.plugins.suppliestracker.SuppliesTrackerPlugin.lambda$buildEntries$0(SuppliesTrackerPlugin.java:466)
        at java.awt.event.InvocationEvent.dispatch(Unknown Source)
        at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
        at java.awt.EventQueue.access$500(Unknown Source)
        at java.awt.EventQueue$3.run(Unknown Source)
        at java.awt.EventQueue$3.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
        at java.awt.EventQueue.dispatchEvent(Unknown Source)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
        at java.awt.EventDispatchThread.run(Unknown Source)`

**Comment by james-munson** [comment]
```
2019-03-11 00:59:49 [AWT-EventQueue-0] ERROR net.runelite.client.RuneLite - Uncaught exception:
java.util.ConcurrentModificationException: null
        at java.util.HashMap$HashIterator.nextNode(Unknown Source)
        at java.util.HashMap$KeyIterator.next(Unknown Source)
        at net.runelite.client.plugins.suppliestracker.SuppliesTrackerPanel.addRow(SuppliesTrackerPanel.java:361)
        at net.runelite.client.plugins.suppliestracker.SuppliesTrackerPlugin.lambda$buildEntries$0(SuppliesTrackerPlugin.java:466)
        at java.awt.event.InvocationEvent.dispatch(Unknown Source)
        at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
        at java.awt.EventQueue.access$500(Unknown Source)
        at java.awt.EventQueue$3.run(Unknown Source)
        at java.awt.EventQueue$3.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
        at java.awt.EventQueue.dispatchEvent(Unknown Source)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
        at java.awt.EventDispatchThread.run(Unknown Source)
```
though the supply tracker is looking nice
it tracks the charges on the trident but im almost pretty sure that its the trident which is causing this issue

File: runelite-client/src/main/java/net/runelite/client/plugins/suppliestracker/SuppliesTrackerPlugin.java [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+		clientToolbar.removeNavigation(navButton);
+	}
+
+	@Subscribe
+	public void onAnimationChanged(AnimationChanged animationChanged) throws ExecutionException
+	{
+
+		if (animationChanged.getActor() == client.getLocalPlayer())
+		{
+			if (animationChanged.getActor().getAnimation() == 1167)
```
Comment (daviscook477) [inline runelite-client/src/main/java/net/runelite/client/plugins/suppliestracker/SuppliesTrackerPlugin.java]: this constant 1167 should be added as a constant to net.runelite.api.AnimationID and then imported into this plugin

**Comment by daviscook477** [comment]
@Beagon 

is this more in line with what you were thinking for a supply tracker ui?
suggested_ui

**Comment by daviscook477** [comment]
so what supplies would need to be tracked in order to consider an implementation of the supply tracker complete?

@Dyldozer has code to track "consumables, ammo, trident casts, teleport tablets, and teleport scrolls" however there are a lot more things that could be considered supplies that are used on pvm trips/boss fights/slayer trasks/etc...

@leele2 mentions the blowpipe and enchanted jewlery

and runes for spellcasts like high alchemy, vengeance, etc... seems important.

**Comment by daviscook477** [comment]
on topic of blowpipe: the blowpipe doesn't have client side the count of its darts/scales (https://github.com/runelite/runelite/issues/953) but the scales can be tracked by animation changes to blowpipe attack. but we never have dart type information (even after checking the varbits) so we could mock a "dart" and then just add .2 or .28 or 1 darts per animation to estimate the avas effect. price calc wouldn't work but there could be an option in the plugin's config for what dart type to count the blowpipe as since it's such a common weapon.

**Pull request closed without merging**